src/mon/PGMap.cc

```diff
@@ -539,6 +539,13 @@ void PGMap::stat_pg_update(const pg_t pgid, pg_stat_t& s,
     s.blocked_by == n.blocked_by;
 
   stat_pg_sub(pgid, s, sameosds);
+
+  // if acting_primary has shift to an just restored osd, and pg yet to finish
+  // peering, many attributes in current stats remain stale. others seem don't
+  // mater much while faulty last_active will make "pg stuck in" check unhappy.
+  if ((s.acting_primary != n.acting_primary) &&
+      !(n.state & (PG_STATE_ACTIVE | PG_STATE_PEERED)))
+    n.last_active = s.last_active;
```

Comment:
yep, changing last_active happens at publish_stats_to_osd() if state is PG_STATE_ACTIVE or at PG creating. but, suppose that, primary OSD was down for some reason,PG should peering -> active, continuously updating last_active by new primary, then the old primary OSD restored within few minutes,whose pglog catch up the tail, this OSD will be selected as primary again soon, before its statemachine goes to GotInfo, its pgstats remain stale, these stats may be sent to Monitor and misleading it, so we should handle this.